pkg/services/grafana-apiserver/utils/meta.go

```diff
@@ -0,0 +1,266 @@
+
+	// Find a title in the object
+	// This will reflect the object and try to get:
+	//  * spec.title
+	//  * spec.name
+	//  * title
+	// and return an empty string if nothing was found
+	FindTitle(defaultTitle string) string
+}
+
+var _ GrafanaResourceMetaAccessor = (*grafanaResourceMetaAccessor)(nil)
+
+type grafanaResourceMetaAccessor struct {
+	raw interface{} // the original object (it implements metav1.Object)
+	obj metav1.Object
+}
+
+// Accessor takes an arbitrary object pointer and returns meta.Interface.
+// obj must be a pointer to an API type. An error is returned if the minimum
+// required fields are missing. Fields that are not required return the default
+// value and are a no-op if set.
+func MetaAccessor(raw interface{}) (GrafanaResourceMetaAccessor, error) {
+	obj, err := meta.Accessor(raw)
+	if err != nil {
+		return nil, err
+	}
+	return &grafanaResourceMetaAccessor{raw, obj}, nil
+}
+
+func (m *grafanaResourceMetaAccessor) set(key string, val string) {
+	anno := m.obj.GetAnnotations()
+	if val == "" {
+		if anno != nil {
+			delete(anno, key)
+		}
+	} else {
+		if anno == nil {
+			anno = make(map[string]string)
+		}
+		anno[key] = val
+	}
+	m.obj.SetAnnotations(anno)
+}
+
+func (m *grafanaResourceMetaAccessor) get(key string) string {
+	return m.obj.GetAnnotations()[key]
+}
+
+func (m *grafanaResourceMetaAccessor) GetUpdatedTimestamp() (*time.Time, error) {
+	v, ok := m.obj.GetAnnotations()[AnnoKeyUpdatedTimestamp]
+	if !ok || v == "" {
+		return nil, nil
+	}
+	t, err := time.Parse(time.RFC3339, v)
+	if err != nil {
+		return nil, fmt.Errorf("invalid updated timestamp: %s", err.Error())
+	}
+	return &t, nil
+}
+
+func (m *grafanaResourceMetaAccessor) SetUpdatedTimestampMillis(v int64) {
+	if v > 0 {
+		t := time.UnixMilli(v)
+		m.SetUpdatedTimestamp(&t)
+	} else {
+		m.set(AnnoKeyUpdatedTimestamp, "") // will clear the annotation
+	}
+}
+
+func (m *grafanaResourceMetaAccessor) SetUpdatedTimestamp(v *time.Time) {
+	txt := ""
+	if v != nil && v.Unix() != 0 {
+		txt = v.UTC().Format(time.RFC3339)
+	}
+	m.set(AnnoKeyUpdatedTimestamp, txt)
+}
+
+func (m *grafanaResourceMetaAccessor) GetCreatedBy() string {
+	return m.get(AnnoKeyCreatedBy)
+}
+
+func (m *grafanaResourceMetaAccessor) SetCreatedBy(user string) {
+	m.set(AnnoKeyCreatedBy, user)
+}
+
+func (m *grafanaResourceMetaAccessor) GetUpdatedBy() string {
+	return m.get(AnnoKeyUpdatedBy)
+}
+
+func (m *grafanaResourceMetaAccessor) SetUpdatedBy(user string) {
+	m.set(AnnoKeyUpdatedBy, user)
+}
+
+func (m *grafanaResourceMetaAccessor) GetFolder() string {
+	return m.get(AnnoKeyFolder)
+}
+
+func (m *grafanaResourceMetaAccessor) SetFolder(uid string) {
+	m.set(AnnoKeyFolder, uid)
+}
+
+func (m *grafanaResourceMetaAccessor) GetSlug() string {
+	return m.get(AnnoKeySlug)
+}
+
+func (m *grafanaResourceMetaAccessor) SetSlug(v string) {
+	m.set(AnnoKeySlug, v)
+}
+
+func (m *grafanaResourceMetaAccessor) SetOriginInfo(info *ResourceOriginInfo) {
```

Comment:
nit, but this one might be easier to read and test if it were split into `clearOriginAnnotation` and `SetOriginInfo`